### PR TITLE
sdig: add DoT

### PR DESCRIPTION
**Pull request by omoerbeek**
### Short description

Add client part to `tcpiohandler`, use it in `sdig` to provide DoT.

`tcpiohandler` client parts mostly taken from a branch by @rgacogne;  a few few fields for client specific data were added to existing classes, specifically to store the client context and server hostname.

For the rest the existing server side code should not have been touched, except for a case where an more elaborate error message is constructed in verbose mode, see `convertIORequestToIOState()`

This code is also suitable for rec to get DoT to auth.

### Checklist

I have:
- [X] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [X] compiled this code
- [X] tested this code
- [X] included documentation (including possible behaviour changes)
- [X] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

File: pdns/tcpiohandler.cc [hunk 1]
```diff
@@ -0,0 +1,1269 @@
...
+    }
+    const auto ocspMap = reinterpret_cast<std::map<int, std::string>*>(arg);
+    return libssl_ocsp_stapling_callback(ssl, *ocspMap);
+  }
+
+  std::unique_ptr<TLSConnection> getConnection(int socket, unsigned int timeout, time_t now) override
+  {
+    handleTicketsKeyRotation(now);
+
+    return std::unique_ptr<OpenSSLTLSConnection>(new OpenSSLTLSConnection(socket, timeout, d_feContext));
```
Comment (rgacogne) [inline pdns/tcpiohandler.cc:529]: We have C++17 now :)
```suggestion
    return std::make_unique_ptr<OpenSSLTLSConnection>(socket, timeout, d_feContext);
```

File: pdns/tcpiohandler.cc [hunk 2]
```diff
@@ -0,0 +1,1269 @@
...
+  }
+
+  std::unique_ptr<TLSConnection> getClientConnection(const std::string& host, int socket, unsigned int timeout) override
+  {
+    return std::unique_ptr<OpenSSLTLSConnection>(new OpenSSLTLSConnection(host, socket, timeout, d_tlsCtx.get()));
```
Comment (rgacogne) [inline pdns/tcpiohandler.cc:534]:
```suggestion
    return std::make_unique_ptr<OpenSSLTLSConnection>(host, socket, timeout, d_tlsCtx.get());
```

File: pdns/tcpiohandler.cc [hunk 3]
```diff
@@ -0,0 +1,1269 @@
...
+  {
+    handleTicketsKeyRotation(now);
+
+    std::shared_ptr<GnuTLSTicketsKey> ticketsKey;
+    {
+      ReadLock rl(&d_lock);
+      ticketsKey = d_ticketsKey;
+    }
+
+    return std::unique_ptr<GnuTLSConnection>(new GnuTLSConnection(socket, timeout, d_creds.get(), d_priorityCache, ticketsKey, d_enableTickets));
```
Comment (rgacogne) [inline pdns/tcpiohandler.cc:1151]:
```suggestion
    return std::make_unique_ptr<GnuTLSConnection>(socket, timeout, d_creds.get(), d_priorityCache, ticketsKey, d_enableTickets);
```

File: pdns/tcpiohandler.cc [hunk 4]
```diff
@@ -0,0 +1,1269 @@
...
+  }
+
+  std::unique_ptr<TLSConnection> getClientConnection(const std::string& host, int socket, unsigned int timeout) override
+  {
+    return std::unique_ptr<GnuTLSConnection>(new GnuTLSConnection(host, socket, timeout, d_creds.get(), d_priorityCache, d_validateCerts));
```
Comment (rgacogne) [inline pdns/tcpiohandler.cc:1156]:
```suggestion
    return std::make_unique_ptr<GnuTLSConnection>(host, socket, timeout, d_creds.get(), d_priorityCache, d_validateCerts);
```

**Review by rgacogne — APPROVED** [review]
Looks great, well done!

**Comment by omoerbeek** [comment]
On CentOS builld fails:

```
tcpiohandler.cc:738:7: error: 'gnutls_session_set_verify_cert' was not declared in this scope
       gnutls_session_set_verify_cert(d_conn.get(), d_host.c_str(), GNUTLS_VERIFY_ALLOW_UNSORTED_CHAIN);
```
GnuTLS is old there. Solve like OpenSSL, which does not do validation in the corresponding case?

Looking at the code,I wonder if it should be conditional on `validateCerts` and not on `!d_host.empty()`, as well, as the OpenSSL code does. It would be nice if the providers agree....

File: m4/pdns_with_gnutls.m4 [hunk 5]
```diff
@@ -18,7 +18,7 @@ AC_DEFUN([PDNS_WITH_GNUTLS], [
         save_LIBS=$LIBS
         CFLAGS="$GNUTLS_CFLAGS $CFLAGS"
         LIBS="$GNUTLS_LIBS $LIBS"
-        AC_CHECK_FUNCS([gnutls_memset])
+        AC_CHECK_FUNCS([gnutls_memset, gnutls_session_set_verify_cert])
```
Comment (rgacogne) [inline m4/pdns_with_gnutls.m4:21]: We don't usually use commas between the functions, does that work?

File: docs/manpages/sdig.1.rst [hunk 6]
```diff
@@ -44,7 +44,9 @@ dot
 insecure
     when using DoT, do not validate the server certificate.
 subjectName *name*
-    when using DoT, verify the server certificate is issued for *name*.
+    when using DoT, verify the server certificate is issued for *name*. The `openssl` provider will accept an empty name and still
+    make sure the certificate is issued by a trusted CA, `gnutls` will only do the validation if a name is given.
+    Default is the empty name.
```
Comment (rgacogne) [inline docs/manpages/sdig.1.rst:49]: Perhaps we should document that certificate validation is not done in some old versions?

File: docs/manpages/sdig.1.rst [hunk 7]
```diff
@@ -39,6 +39,18 @@ showflags
...
+dot
+    use DoT instead of UDP to send a query. Implies tcp.
+insecure
+    when using DoT, do not validate the server certificate.
+subjectName *name*
+    when using DoT, verify the server certificate is issued for *name*. The `openssl` provider will accept an empty name and still
+    make sure the certificate is issued by a trusted CA, `gnutls` will only do the validation if a name is given.
+    Default is the empty name.
+caStore *file*
+    when using Dot, read the trusted CA certificates from *file*. Default is to use the system provided CA store.
```
Comment (jsoref) [inline docs/manpages/sdig.1.rst:51]:
```suggestion
    when using DoT, read the trusted CA certificates from *file*. Default is to use the system provided CA store.
```

File: pdns/sdig.cc [hunk 8]
```diff
@@ -195,18 +202,26 @@ static void printReply(const string& reply, bool showflags, bool hidesoadetails)
 
 int main(int argc, char** argv)
 try {
+    /* default timeout of 10s */
```
Comment (jsoref) [inline pdns/sdig.cc:205]:
```suggestion
  /* default timeout of 10s */
```

File: pdns/sdig.cc [hunk 9]
```diff
@@ -259,14 +278,35 @@ try {
         xpfsrc = argv[++i];
         xpfdst = argv[++i];
       }
-      if (strcmp(argv[i], "class") == 0) {
+      else if (strcmp(argv[i], "class") == 0) {
         if (argc < i+2) {
           cerr << "class needs an argument"<<endl;
```
Comment (jsoref) [inline pdns/sdig.cc:283]: I know this isn't you, but mostly they do this
```suggestion
          cerr << "class needs an argument" << endl;
```

File: pdns/sdig.cc [hunk 10]
```diff
@@ -259,14 +278,35 @@ try {
...
           exit(EXIT_FAILURE);
         }
         qclass = atoi(argv[++i]);
       }
-      if (strcmp(argv[i], "proxy") == 0) {
+      else if (strcmp(argv[i], "subjectName") == 0) {
+        if (argc < i + 2) {
+          cerr << "subjectName needs an argument"<<endl;
```
Comment (jsoref) [inline pdns/sdig.cc:290]:
```suggestion
          cerr << "subjectName needs an argument" << endl;
```

File: pdns/sdig.cc [hunk 11]
```diff
@@ -259,14 +278,35 @@ try {
...
+          exit(EXIT_FAILURE);
+        }
+        subjectName = argv[++i];
+      }
+      else if (strcmp(argv[i], "caStore") == 0) {
+        if (argc < i + 2) {
+          cerr << "caStore needs an argument"<<endl;
```
Comment (jsoref) [inline pdns/sdig.cc:297]:
```suggestion
          cerr << "caStore needs an argument" << endl;
```

File: pdns/sdig.cc [hunk 12]
```diff
@@ -259,14 +278,35 @@ try {
...
+          exit(EXIT_FAILURE);
+        }
+        caStore = argv[++i];
+      }
+      else if (strcmp(argv[i], "tlsProvider") == 0) {
+        if (argc < i + 2) {
+          cerr << "tlsProvider needs an argument"<<endl;
```
Comment (jsoref) [inline pdns/sdig.cc:304]:
```suggestion
          cerr << "tlsProvider needs an argument" << endl;
```

File: m4/pdns_with_gnutls.m4 (hunk 5 above)
Comment (omoerbeek) [reply m4/pdns_with_gnutls.m4:21]: it does work, but fixing to conform anyway

File: pdns/sdig.cc [hunk 13]
```diff
@@ -259,14 +278,35 @@ try {
...
+          exit(EXIT_FAILURE);
+        }
+        tlsProvider = argv[++i];
+      }
+      else if (strcmp(argv[i], "proxy") == 0) {
         if(argc < i+4) {
           cerr<<"proxy needs three arguments"<<endl;
```
Comment (jsoref) [inline pdns/sdig.cc:311]: not you...
```suggestion
          cerr << "proxy needs three arguments" << endl;
```

File: pdns/sdig.cc [hunk 14]
```diff
@@ -276,9 +316,24 @@ try {
         ComboAddress dest(argv[++i]);
         proxyheader = makeProxyHeader(ptcp, src, dest, {});
       }
+      else {
+        cerr << argv[i] << ": unknown argument" << endl;
```
Comment (jsoref) [inline pdns/sdig.cc:320]: 👍

File: pdns/tcpiohandler.hh [hunk 15]
```diff
@@ -342,4 +375,17 @@ public:
...
+{
+  std::string d_provider;
+  std::string d_ciphers;
+  std::string d_ciphers13;
+  std::string d_caStore;
+  bool d_validateCertificates{true};
+};
+
+std::shared_ptr<TLSCtx> getTLSContext(const TLSContextParameters& params);
+
```
Comment (jsoref) [inline pdns/tcpiohandler.hh:391]:
```suggestion

```

**Comment by omoerbeek** [comment]
@jsoref thanks for the commens, I did not take all white-space ones. These will be done when we get to enforcing formatting in more parts of the tree.

**Pull request merged**